File: c++/determinacao.cpp
#include <bits/stdc++.h>

using namespace std;

#define Matriz vector<vector<double>>

Matriz A, G, Gt;
int n;

Matriz calculo_G();
Matriz transposta(Matriz);
void imprimir(Matriz);

int main(){
    cin >> n;

    A.resize(n, vector<double>(n));
    G.resize(n, vector<double>(n, 0));

    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            cin >> A[i][j];

    cout << "Matriz A:" << endl;
    imprimir(A);
    cout << endl;

    //Calculo da matriz G
    G = calculo_G();
    
    //Determinação de Gt
    Gt = transposta(G);

    //Impressão dos resultados
    cout << "Matriz G:" << endl;
    imprimir(G);
    cout << endl;

    cout << "Matriz Gt:" << endl;
    imprimir(Gt);
    cout << endl;

    return 0;
}

//Determina G
Matriz calculo_G(){
    Matriz mat(n, vector<double>(n, 0));
    
    for(int j = 0; j < n; j++){
        for(int i = j; i < n; i++){
            if(i == 0 && j == 0){
                mat[0][0] = sqrt(A[0][0]);
            }
            if(i == j){
                double soma = 0;

                for(int k = 0; k < i; k++)
                    soma += pow(mat[i][k], 2);

                mat[i][i] = sqrt(A[i][i] - soma);
            }
            if(j == 0){
                mat[i][0] = A[i][0]/mat[0][0];
            }
            else{
                double soma = 0;

                for(int k = 0; k < j; k++)
                    soma += mat[i][k]*mat[j][k];

                mat[i][j] = (A[i][j] - soma)/mat[j][j];

            } 
        }
    }

    return mat;
}

//Imprime uma matriz
void imprimir(Matriz mat){
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++)
            printf("%2.0f ", mat[i][j]);
        cout << endl;
    }
}

//Retorna a transposta de uma matriz
Matriz transposta(Matriz mat){
    Matriz T(n, vector<double>(n));

    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            T[i][j] = mat[j][i];

    return T;
}

File: python/determinacao.py
import numpy as np
import math

n = int(input())
A = np.zeros((n, n))

for i in range(n):
    A[i] = list(map(float, input().split()))

def imprimir(mat):
    for i in mat:
        print(i)
    print()

def calculo_G(mat): 
    G = np.zeros((n, n))

    for j in range(n):
        for i in range(j, n):
            if i == 0 and j == 0:
                G[0][0] = math.sqrt(A[0][0])
            if i == j:
                soma = sum(G[i][k]**2 for k in range(i))
                G[i][i] = math.sqrt(A[i][i] - soma)
            elif j == 0:
                G[i][0] = A[i][0] / G[0][0]
            else:
                soma = sum(G[i][k] * G[j][k] for k in range(j))
                G[i][j] = (A[i][j] - soma) / G[j][j]

    return G

G = calculo_G(A)
Gt = G.T

print("\nA:")
imprimir(A)
print("G:")
imprimir(G)
print("Gt:")
imprimir(Gt)
